R/touchstone_demographic_dataset.R

```diff
@@ -0,0 +1,245 @@
+
+      tdd_touchstones = db_get(con, "touchstone", "id",
+        unique(e_tdd_csv$touchstone))
+    ))
+
+  # Now the demographic_dataset (which is source, type). For the sake of
+  # simplicity, look for matches of either source, or type, rather than
+  # filtering to the exact combinations, which would be a bit complicated
+  # at this point, and not worth much. There aren't very many of these.
+
+    e <- c(e, list(db_dd = db_dd_info(con,
+      e_tdd_csv$demographic_statistic_type,
+      e_tdd_csv$demographic_source)))
+
+  # In the extract tests and the transform, I need to work with the demographic
+  # dataset, so will mash them together once here. Not exactly an extract, but
+  # a very simple thing that saves some repetition later.
+
+    # A mash to compare with existing touchstone_demographic_datasets
+
+    e$touchstone_demographic_dataset_csv$mash <- paste(
+      e_tdd_csv$touchstone, e_tdd_csv$demographic_source,
+      e_tdd_csv$demographic_statistic_type, sep = '\r')
+
+    e$db_tdd$mash <- paste(e$db_tdd$touchstone,
+                           e$db_tdd$dsource_code,
+                           e$db_tdd$dtype_code, sep = '\r')
+
+    # A mash to compare with existing demographic_datasets
+
+    e$touchstone_demographic_dataset_csv$ds_mash <- paste(
+      e_tdd_csv$demographic_source,
+      e_tdd_csv$demographic_statistic_type,
+      sep = '\r')
+
+    e$db_dd$mash <- paste(e$db_dd$dsource_code,
+                          e$db_dd$dtype_code, sep = '\r')
+    e
+
+  } else {
+    list()
+  }
+}
+
+test_extract_touchstone_demographic_dataset <- function(e) {
+  if (!is.null(e$touchstone_demographic_dataset_csv)) {
+    e_tdd_csv <- e$touchstone_demographic_dataset_csv
+
+    testthat::expect_true(all(e_tdd_csv$demographic_source %in%
+                              e$db_dsrc$code),
+      label = "demographic sources in touchstone_demographic_dataset exist")
+
+    testthat::expect_true(all(e_tdd_csv$demographic_statistic_type %in%
+                    e$db_dstype$code),
+      label = "demog. statistic types in touchstone_demographic_dataset exist")
+
+    all_touchstones <- unique(c(e$tdd_touchstones$id, e$touchstone_csv$id))
+
+    testthat::expect_true(all(e_tdd_csv$touchstone %in% all_touchstones),
+      label = "Touchstones in touchstone_demographic_dataset exist")
+
+    testthat::expect_true(all(e_tdd_csv$ds_mash %in% e$db_dd$mash),
+      label = "Demographic datasets already exist")
+
+    testthat::expect_equal(0, sum(duplicated(e_tdd_csv$mash)),
+      label = "No duplicated (Touchstone, demographic_dataset) pair in csv")
+  }
+}
+
+###############################################################################
+
+transform_touchstone_demographic_dataset <- function(e) {
+
+  tdd <- e$touchstone_demographic_dataset_csv
+
+  if (!is.null(tdd)) {
+    tdd$demographic_dataset <- e$db_dd$id[match(tdd$ds_mash, e$db_dd$mash)]
+
+  # If any (touchstone,dataset) already exist in db's
```

Comment:
indenting needs redoing

Reply:
Done - as product of the below change...